library/std/src/collections/hash/map.rs

```diff
@@ -1151,6 +1154,43 @@ where
     }
 }
 
+#[stable(feature = "std_collections_from_array", since = "1.53.0")]
+// Note: as what is currently the most convenient built-in way to construct
+// a HashMap, a simple usage of this function must not *require* the user
+// to provide a type annotation in order to infer the third type parameter
+// (the hasher parameter, conventionally "S").
+// To that end, this impl is defined using RandomState as the concrete
+// type of S, rather than being generic over `S: BuildHasher + Default`.
+// It is expected that users who want to specify a hasher will manually use
+// `with_capacity_and_hasher`.
+// If type parameter defaults worked on impls, and if type parameter
+// defaults could be mixed with const generics, then perhaps
+// this could be generalized.
+// See also the equivalent impl on HashSet.
+impl<K, V, const N: usize> From<[(K, V); N]> for HashMap<K, V, RandomState>
+where
+    K: Eq + Hash,
+{
+    /// ```
+    /// use std::collections::HashMap;
+    ///
+    /// let map1 = HashMap::from([(1, 2), (3, 4)]);
+    /// let map2: HashMap<_, _> = [(1, 2), (3, 4)].into();
+    /// assert_eq!(map1, map2);
+    ///
+    /// let unordered_duplicates = HashMap::from([(3, 4), (1, 2), (1, 2)]);
+    /// assert_eq!(map1, unordered_duplicates);
+    ///
+    /// // This next line must infer the hasher type parameter.
+    /// // If you make a change that causes this line to no longer infer,
+    /// // that's a problem!
+    /// let must_not_require_type_annotation = HashMap::from([(1, 2)]);
```

Comment:
Can we put this to unit test rather than showing it in the docs?

Reply:
Agreed, this doesn't belong in the documentation.

Reply:
I wasn't aware that docs on impls actually got rendered by rustdoc, though perhaps the reason I've never seen it is simply because it's rare for anybody to actually do it. I was just using the doctest as a lazy excuse to not write a unit test. :P I'll go ahead and make these into proper unit tests.